File: typescript/packages/server/src/routers/workspaces.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { mkdtempSync, readFileSync, rmSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import type { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify'
import type { MountMode } from '@struktoai/mirage-node'
import { Workspace, type Resource } from '@struktoai/mirage-node'
import type { WorkspaceRegistry } from '../registry.ts'
import { buildOverrideResources, cloneWorkspaceWithOverride, type OverrideShape } from '../clone.ts'
import {
  configToWorkspaceArgs,
  loadWorkspaceConfig,
  type WorkspaceArgs,
  type WorkspaceConfigRaw,
} from '../config.ts'
import { makeBrief, makeDetail } from '../summary.ts'

export interface WorkspaceRoutesDeps {
  registry: WorkspaceRegistry
}

interface CreateWorkspaceBody {
  config: Record<string, unknown>
  id?: string
}

interface WorkspaceIdParams {
  id: string
}

interface WorkspaceGetQuery {
  verbose?: string
}

interface CloneWorkspaceBody {
  id?: string
  override?: OverrideShape
}

export function registerWorkspacesRoutes(app: FastifyInstance, deps: WorkspaceRoutesDeps): void {
  app.post<{ Body: CreateWorkspaceBody }>(
    '/v1/workspaces',
    async (req: FastifyRequest<{ Body: CreateWorkspaceBody }>, reply: FastifyReply) => {
      const body = req.body
      if (body.id !== undefined && deps.registry.has(body.id)) {
        return reply.status(409).send({ detail: `workspace id already exists: ${body.id}` })
      }
      let cfg: WorkspaceConfigRaw
      try {
        cfg = loadWorkspaceConfig(body.config)
      } catch (e) {
        return reply.status(400).send({ detail: (e as Error).message })
      }
      let args: WorkspaceArgs
      try {
        args = await configToWorkspaceArgs(cfg)
      } catch (e) {
        return reply.status(502).send({ detail: `resource build failed: ${(e as Error).message}` })
      }
      const resourceMap: Record<string, Resource> = {}
      const modeOverrides: Record<string, MountMode> = {}
      for (const [prefix, [resource, mode]] of Object.entries(args.resources)) {
        resourceMap[prefix] = resource
        modeOverrides[prefix] = mode
      }
      const ws = new Workspace(resourceMap, {
        mode: args.options.mode,
        modeOverrides,
        ...(args.options.cache !== undefined ? { cache: args.options.cache } : {}),
        ...(args.options.index !== undefined ? { index: args.options.index } : {}),
      })
      let entry
      try {
        entry = deps.registry.add(ws, body.id)
      } catch (e) {
        return reply.status(409).send({ detail: (e as Error).message })
      }
      return reply.status(201).send(makeDetail(entry))
    },
  )

  app.get('/v1/workspaces', () => deps.registry.list().map(makeBrief))

  app.post('/v1/workspaces/load', async (req, reply) => {
    let tarBuf: Buffer | null = null
    let workspaceId: string | undefined
    let override: unknown = null
    for await (const part of req.parts()) {
      if (part.type === 'file' && part.fieldname === 'tar') {
        tarBuf = await part.toBuffer()
      } else if (part.type === 'field' && part.fieldname === 'id') {
        workspaceId = String(part.value)
      } else if (part.type === 'field' && part.fieldname === 'override') {
        try {
          override = JSON.parse(String(part.value))
        } catch {
          return reply.status(400).send({ detail: 'override must be JSON' })
        }
      }
    }
    if (tarBuf === null) return reply.status(400).send({ detail: 'missing tar field' })
    if (workspaceId !== undefined && deps.registry.has(workspaceId)) {
      return reply.status(409).send({ detail: `workspace id already exists: ${workspaceId}` })
    }
    let overrides: Record<string, Resource>
    try {
      overrides = await buildOverrideResources(override as OverrideShape | null)
    } catch (e) {
      return reply.status(400).send({ detail: `override build failed: ${(e as Error).message}` })
    }
    let ws: Workspace
    try {
      ws = await Workspace.load(new Uint8Array(tarBuf), {}, overrides)
    } catch (e) {
      return reply.status(400).send({ detail: `load failed: ${(e as Error).message}` })
    }
    let entry
    try {
      entry = deps.registry.add(ws, workspaceId)
    } catch (e) {
      return reply.status(409).send({ detail: (e as Error).message })
    }
    return reply.status(201).send(makeDetail(entry))
  })

  app.get<{ Params: WorkspaceIdParams; Querystring: WorkspaceGetQuery }>(
    '/v1/workspaces/:id',
    (req, reply) => {
      const { id } = req.params
      if (!deps.registry.has(id)) return reply.status(404).send({ detail: 'workspace not found' })
      const verbose = req.query.verbose === 'true'
      return makeDetail(deps.registry.get(id), verbose)
    },
  )

  app.delete<{ Params: WorkspaceIdParams }>('/v1/workspaces/:id', async (req, reply) => {
    const { id } = req.params
    if (!deps.registry.has(id)) return reply.status(404).send({ detail: 'workspace not found' })
    await deps.registry.remove(id)
    return { id, closedAt: Date.now() / 1000 }
  })

  app.post<{ Params: WorkspaceIdParams; Body: CloneWorkspaceBody }>(
    '/v1/workspaces/:id/clone',
    async (req, reply) => {
      const { id } = req.params
      if (!deps.registry.has(id)) return reply.status(404).send({ detail: 'workspace not found' })
      const body = req.body
      if (body.id !== undefined && deps.registry.has(body.id)) {
        return reply.status(409).send({ detail: `workspace id already exists: ${body.id}` })
      }
      const src = deps.registry.get(id).runner.ws
      const newWs = await cloneWorkspaceWithOverride(src, body.override ?? null)
      let entry
      try {
        entry = deps.registry.add(newWs, body.id)
      } catch (e) {
        return reply.status(409).send({ detail: (e as Error).message })
      }
      return reply.status(201).send(makeDetail(entry))
    },
  )

  app.get<{ Params: WorkspaceIdParams }>('/v1/workspaces/:id/snapshot', async (req, reply) => {
    const { id } = req.params
    if (!deps.registry.has(id)) return reply.status(404).send({ detail: 'workspace not found' })
    const tmp = mkdtempSync(join(tmpdir(), 'mirage-snap-'))
    const out = join(tmp, `${id}.tar`)
    try {
      await deps.registry.get(id).runner.ws.snapshot(out)
      const buf = readFileSync(out)
      reply.header('Content-Type', 'application/x-tar')
      reply.header('Content-Disposition', `attachment; filename="${id}.tar"`)
      return await reply.send(buf)
    } finally {
      rmSync(tmp, { recursive: true, force: true })
    }
  })
}

File: python/mirage/server/routers/workspaces.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException, Query, Request

from mirage import Workspace
from mirage.resource.registry import build_resource
from mirage.server.clone import clone_workspace_with_override
from mirage.server.summary import make_brief, make_detail
from mirage.workspace.snapshot.utils import norm_mount_prefix

from mirage.server.schemas import (  # isort: skip
    CloneWorkspaceRequest, CreateWorkspaceRequest, DeleteWorkspaceResponse,
    LoadWorkspaceRequest, SnapshotWorkspaceRequest, SnapshotWorkspaceResponse,
    WorkspaceBrief, WorkspaceDetail)

router = APIRouter(prefix="/v1/workspaces")


@router.post("", response_model=WorkspaceDetail, status_code=201)
async def create_workspace(req: CreateWorkspaceRequest,
                           request: Request) -> WorkspaceDetail:
    registry = request.app.state.registry
    if req.id is not None and req.id in registry:
        raise HTTPException(status_code=409,
                            detail=f"workspace id already exists: {req.id!r}")
    kwargs = req.config.to_workspace_kwargs()
    ws = Workspace(**kwargs)
    try:
        entry = registry.add(ws, workspace_id=req.id)
    except ValueError as e:
        raise HTTPException(status_code=409, detail=str(e))
    return make_detail(entry)


@router.get("", response_model=list[WorkspaceBrief])
async def list_workspaces(request: Request) -> list[WorkspaceBrief]:
    return [make_brief(e) for e in request.app.state.registry.list()]


@router.get("/{workspace_id}", response_model=WorkspaceDetail)
async def get_workspace(
    workspace_id: str, request: Request, verbose: bool = Query(False)
) -> WorkspaceDetail:  # noqa: E125
    registry = request.app.state.registry
    if workspace_id not in registry:
        raise HTTPException(status_code=404, detail="workspace not found")
    return make_detail(registry.get(workspace_id), verbose=verbose)


@router.delete("/{workspace_id}", response_model=DeleteWorkspaceResponse)
async def delete_workspace(workspace_id: str,
                           request: Request) -> DeleteWorkspaceResponse:
    import time
    registry = request.app.state.registry
    if workspace_id not in registry:
        raise HTTPException(status_code=404, detail="workspace not found")
    await registry.remove(workspace_id)
    return DeleteWorkspaceResponse(id=workspace_id, closed_at=time.time())


@router.post("/{workspace_id}/clone",
             response_model=WorkspaceDetail,
             status_code=201)
async def clone_workspace(workspace_id: str, req: CloneWorkspaceRequest,
                          request: Request) -> WorkspaceDetail:
    registry = request.app.state.registry
    if workspace_id not in registry:
        raise HTTPException(status_code=404, detail="workspace not found")
    if req.id is not None and req.id in registry:
        raise HTTPException(status_code=409,
                            detail=f"workspace id already exists: {req.id!r}")
    src_entry = registry.get(workspace_id)
    new_ws = await src_entry.runner.call(
        clone_workspace_with_override(src_entry.runner.ws, req.override))
    try:
        entry = registry.add(new_ws, workspace_id=req.id)
    except ValueError as e:
        raise HTTPException(status_code=409, detail=str(e))
    return make_detail(entry)


@router.post("/{workspace_id}/snapshot",
             response_model=SnapshotWorkspaceResponse)
async def snapshot_workspace(workspace_id: str, req: SnapshotWorkspaceRequest,
                             request: Request) -> SnapshotWorkspaceResponse:
    registry = request.app.state.registry
    if workspace_id not in registry:
        raise HTTPException(status_code=404, detail="workspace not found")
    entry = registry.get(workspace_id)
    target = Path(req.path)
    target.parent.mkdir(parents=True, exist_ok=True)
    await entry.runner.call(_run_snapshot(entry.runner.ws, str(target)))
    return SnapshotWorkspaceResponse(id=workspace_id,
                                     path=str(target),
                                     size=target.stat().st_size)


async def _run_snapshot(ws: Workspace, target: str) -> None:
    await ws.snapshot(target)


@router.post("/load", response_model=WorkspaceDetail, status_code=201)
async def load_workspace(req: LoadWorkspaceRequest,
                         request: Request) -> WorkspaceDetail:
    registry = request.app.state.registry
    if req.id is not None and req.id in registry:
        raise HTTPException(status_code=409,
                            detail=f"workspace id already exists: {req.id!r}")
    resources = _build_load_resources(req.override)
    try:
        ws = Workspace.load(req.path, resources=resources)
    except FileNotFoundError:
        raise HTTPException(status_code=400,
                            detail=f"snapshot not found: {req.path}")
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    try:
        entry = registry.add(ws, workspace_id=req.id)
    except ValueError as e:
        raise HTTPException(status_code=409, detail=str(e))
    return make_detail(entry)


def _build_load_resources(override: dict[str, Any] | None) -> dict | None:
    if not override or "mounts" not in override:
        return None
    out: dict = {}
    for prefix, block in override["mounts"].items():
        if not isinstance(block, dict):
            continue
        resource_name = block.get("resource")
        config = block.get("config") or {}
        if resource_name is None:
            continue
        out[norm_mount_prefix(prefix)] = build_resource(resource_name, config)
    return out or None
